src/metabase/api/dashboard.clj

```diff
@@ -683,40 +635,32 @@
-                current-cards             (cond->> (:ordered_cards dashboard)
-                                            (seq deleted-tab-ids)
-                                            (remove (fn [card]
-                                                      (contains? deleted-tab-ids (:dashboard_tab_id card)))))
-                new-cards                 (cond->> cards
-                                            ;; fixup the temporary tab ids with the real ones
-                                            (seq temp->real-tab-ids)
-                                            (map (fn [card]
-                                                   (if-let [real-tab-id (get temp->real-tab-ids (:dashboard_tab_id card))]
-                                                     (assoc card :dashboard_tab_id real-tab-id)
-                                                     card))))]
-            (do-update-dashcards! dashboard current-cards new-cards)
-            (reset! tabs-stats-atom (select-keys tab-stats [:num-tabs-created :num-tabs-deleted :total-num-tabs]))))
-        (let [{:keys [num-tabs-created num-tabs-deleted total-num-tabs]} @tabs-stats-atom]
-          (when (pos-int? num-tabs-created)
-            (snowplow/track-event! ::snowplow/dashboard-tabs-created
-                                   api/*current-user-id*
-                                   {:dashboard-id  (:id dashboard)
-                                    :num-tabs       num-tabs-created
-                                    :total-num-tabs total-num-tabs}))
-          (when (pos-int? num-tabs-deleted)
-            (snowplow/track-event! ::snowplow/dashboard-tabs-deleted
-                                   api/*current-user-id*
-                                   {:dashboard-id  (:id dashboard)
-                                    :num-tabs       num-tabs-deleted
-                                    :total-num-tabs total-num-tabs})))
+                 :as update-tabs-stats}  (dashboard-tab/do-update-tabs! (:id dashboard) (:ordered_tabs dashboard) new-tabs)
+                deleted-tab-ids          (set deleted-tab-ids)
+                current-cards            (cond->> (:ordered_cards dashboard)
+                                           (seq deleted-tab-ids)
+                                           (remove (fn [card]
+                                                     (contains? deleted-tab-ids (:dashboard_tab_id card)))))
+                new-cards                (cond->> cards
+                                           ;; fixup the temporary tab ids with the real ones
+                                           (seq old->new-tab-id)
+                                           (map (fn [card]
+                                                  (if-let [real-tab-id (get old->new-tab-id (:dashboard_tab_id card))]
+                                                    (assoc card :dashboard_tab_id real-tab-id)
+                                                    card))))
```

Comment:
Found a bug. Imagine you start with no tabs, then add some cards to the dash. Save. This is version A. Then you create a tab, save, then delete it, save. So now you have one tab with some cards. Then revert to version A. The one remaining tab should get deleted, but the dashcard’s dashboard_tab_id ID field won’t update, and will continue referencing the deleted tab. You'll need to also fixup dashboard_tab_id for tab IDs that have been deleted.